src/Interpreters/convertFieldToType.cpp

```diff
@@ -109,6 +109,20 @@ Field convertDecimalToDecimalType(const Field & from, const DataTypeDecimal<T> &
     return DecimalField<T>(value, type.getScale());
 }
 
+template <typename From, typename T>
+Field convertFloatToDecimalType(const Field & from, const DataTypeDecimal<T> & type)
```

Comment:
Consider `DecimalUtils::convertTo<Float64>(...)`

Reply:
thank you for advice.
I've used **convertToDecimal** in my func **convertFloatToDecimalType**
that works well

Reply:
@lhuang09287750 Sorry for the delay

The main concern is converting float to string, which looks like a workaround.

What about using a scale of the target type and then checking that we didn't lose any precision (as I did for decimals with different scales)

Here are the changes, you can cherry-pick commits or force-push my branch (https://github.com/ClickHouse/ClickHouse/tree/vdimir/decimal-in-float) into the current PR: https://github.com/ClickHouse/ClickHouse/compare/vdimir/decimal-in-float

Reply:
back from vacation
I've cherry-picked your commits and made a push again!
Your fix also worked,but I don't understand why, what does the function in ActionVistors.cpp do?

Reply:
I make a debug and understand it! :->

Reply:
> what does the function in ActionVistors.cpp do?

It converts decimal back to float and check that it's the same number, so no precision loss.